firebase-firestore/src/main/java/com/google/firebase/firestore/local/BundleSerializer.java

```diff
@@ -0,0 +1,350 @@
+    } else {
+      parent = parent.append(collectionSelector.getString("collectionId"));
+    }
+
+    List<Filter> filters = decodeWhere(structuredQuery.optJSONObject("where"));
+    List<OrderBy> orderBys = decodeOrderBy(structuredQuery.optJSONArray("orderBy"));
+    @Nullable Bound startAt = decodeBound(structuredQuery.optJSONObject("startAt"));
+    @Nullable Bound endAt = decodeBound(structuredQuery.optJSONObject("endAt"));
+
+    verifyNoOffset(structuredQuery);
+    int limit = decodeLimit(structuredQuery);
+    Query.LimitType limitType = decodeLimitType(bundledQueryJson);
+
+    return new Query(parent, collectionGroup, filters, orderBys, limit, limitType, startAt, endAt);
+  }
+
+  private int decodeLimit(JSONObject structuredQuery) {
+    return structuredQuery.optInt("limit", -1);
+  }
+
+  private Bound decodeBound(@Nullable JSONObject bound) throws JSONException {
+    if (bound != null) {
+      List<Value> cursor = new ArrayList<>();
+      boolean before = bound.optBoolean("before", false);
+
+      JSONArray values = bound.optJSONArray("values");
+      if (values != null) {
+        for (int i = 0; i < values.length(); ++i) {
+          cursor.add(decodeValue(values.getJSONObject(i)));
+        }
+      }
+
+      return new Bound(cursor, before);
+    }
+
+    return null;
+  }
+
+  private List<OrderBy> decodeOrderBy(@Nullable JSONArray orderBys) throws JSONException {
+    List<OrderBy> result = new ArrayList<>();
+
+    if (orderBys != null) {
+      for (int i = 0; i < orderBys.length(); ++i) {
+        JSONObject orderBy = orderBys.getJSONObject(i);
+        FieldPath fieldPath = decodeFieldReference(orderBy.getJSONObject("field"));
+        String directionString = orderBy.optString("direction", "ASCENDING");
+        OrderBy.Direction direction =
+            directionString.equals("ASCENDING")
+                ? OrderBy.Direction.ASCENDING
+                : OrderBy.Direction.DESCENDING;
+        result.add(OrderBy.getInstance(direction, fieldPath));
+      }
+    }
+
+    return result;
+  }
+
+  private List<Filter> decodeWhere(@Nullable JSONObject where) throws JSONException {
+    List<Filter> result = new ArrayList<>();
+    if (where != null) {
+      decodeFilter(result, where);
+    }
+    return result;
+  }
+
+  private void decodeFilter(List<Filter> result, JSONObject structuredQuery) throws JSONException {
+    if (structuredQuery.has("compositeFilter")) {
+      decodeCompositeFilter(result, structuredQuery.getJSONObject("compositeFilter"));
+    } else if (structuredQuery.has("fieldFilter")) {
+      decodeFieldFilter(result, structuredQuery.getJSONObject("fieldFilter"));
+    } else if (structuredQuery.has("unaryFilter")) {
+      decodeUnaryFilter(result, structuredQuery.getJSONObject("unaryFilter"));
+    }
+  }
+
+  private void decodeCompositeFilter(List<Filter> result, JSONObject compositeFilter)
+      throws JSONException {
+    JSONArray filters = compositeFilter.optJSONArray("filters");
```

Comment:
Should this verify that the composite filter is AND?

Reply:
We can, but there is no other valid value for now. I added a validation.